### PR TITLE
BUG: setitem casting object Index to arrow strings

### PR DESCRIPTION
**Pull request by phofl**
- [ ] closes #xxxx (Replace xxxx with the GitHub issue number)
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

cc @jbrockmendel 

I think we can treat this as a bug for the new option not a case for deprecation?

I'll open a deprecation PR for all other scenarios after this is in

### CODE REVIEW

**Comment by phofl** [comment]
cc @mroeschke

File: pandas/core/construction.py [hunk 1]
```diff
@@ -548,6 +549,10 @@ def sanitize_array(
         # Avoid ending up with a NumpyExtensionArray
         dtype = dtype.numpy_dtype
 
+    object_index = False
+    if isinstance(data, ABCIndex) and data.dtype == object and dtype is None:
```
Comment (mroeschke) [inline pandas/core/construction.py:553]: We don't have the same issue with `Series[object]`?

File: doc/source/whatsnew/v2.1.2.rst [hunk 2]
```diff
@@ -41,9 +41,11 @@ Bug fixes
 - Fixed bug in :meth:`.DataFrameGroupBy.min()` and :meth:`.DataFrameGroupBy.max()` not preserving extension dtype for empty object (:issue:`55619`)
 - Fixed bug in :meth:`.SeriesGroupBy.value_counts` returning incorrect dtype for string columns (:issue:`55627`)
 - Fixed bug in :meth:`Categorical.equals` if other has arrow backed string dtype (:issue:`55364`)
+- Fixed bug in :meth:`DataFrame.__setitem__` casting :class:`Index` with object-dtype to PyArrow backed strings when ``infer_string`` option is set (:issue:`55638`)
```
Comment (mroeschke) [inline doc/source/whatsnew/v2.1.2.rst:44]: These should go in 2.1.4 now

File: pandas/core/construction.py (hunk 1 above)
Comment (phofl) [reply pandas/core/construction.py:553]: No Series takes an early exit, brock and I chatted about deprecating this behavior for Index altogether as well

File: doc/source/whatsnew/v2.1.2.rst (hunk 2 above)
Comment (phofl) [reply doc/source/whatsnew/v2.1.2.rst]: moved, thx

**Review by mroeschke — COMMENTED** [review]
Looks good. Just a merge conflict

**Review by mroeschke — APPROVED** [review]
cc @jbrockmendel if any other thoughts

**Pull request merged**